🚀  Fast delete cluster resource group when entire cluster is deleted

CecileRobertMichon opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

**What this PR does / why we need it**:  TLDR:
Before:
```
time kubectl delete cluster default-template 
cluster.cluster.x-k8s.io "default-template" deleted
kubectl delete cluster default-template  0.18s user 0.09s system 0% cpu 14:49.77 total
```

After:
```
time kubectl delete cluster default-template              
cluster.cluster.x-k8s.io "default-template" deleted
kubectl delete cluster default-template  0.18s user 0.16s system 0% cpu 8:21.90 total
```

Currently, every Azure resource is deleted serially when a cluster gets deleted. In most cases, it is enough to just delete the resource group. Azure will take care of deleting all resources within that resource group in the proper order.

This does PR attempts to delete the entire resource group and skip deleting individual resources like VMs when 1) the entire cluster is being deleted and 2) when that cluster owns the resource group (which means it gets deleted as part of cluster delete). If these two conditions are not met, normal delete proceeds as before, deleting each resource type by order of dependency.

REVERTED ~This also introduces a "no wait" delete for the resource group delete, which means we don't wait for actual completion. As soon as Azure returns a 200 to confirm that the delete request was received, we return success back to the user and the resource group continues deleting in the background. This is safe for resource groups and resource groups only because it is the last resource type that we delete and no other resources depend on it.~

~What happens if the resource group is still deleting in the background and the user attempts to create a new cluster in that same RG? The user will see errors similar to:~
```
controller-runtime/controller "msg"="Reconciler error" "error"="failed to reconcile cluster services: failed to reconcile virtual network: failed to create virtual network default-template-vnet: network.VirtualNetworksClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=\u003cnil\u003e Code=\"ResourceGroupBeingDeleted\" Message=\"The resource group 'default-template' is in deprovisioning state and cannot perform this operation.\"" "controller"="azurecluster" "name"="default-template" "namespace"="default"
```
~in controller logs until the resource group deletion is completed and the new create can proceed (at which point it does).~

This PR might also help with #937 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

/hold for unit tests

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Directly delete cluster resource group when entire cluster is deleted
```

CecileRobertMichon commented on cloud/services/groups/groups.go:
```diff
@@ -56,18 +56,18 @@ func (s *Service) Reconcile(ctx context.Context) error {
 
 	s.Scope.V(2).Info("deleting resource group", "resource group", s.Scope.ResourceGroup())
-	err = s.Client.Delete(ctx, s.Scope.ResourceGroup())
+	err = s.Client.DeleteNoWait(ctx, s.Scope.ResourceGroup())
```
I can switch this back to Delete() if we prefer to make the RG delete blocking for now

CecileRobertMichon replied:
I measured ~15 minutes to delete the cluster before this change vs. ~7 minutes to delete the entire resource group with this method.

CecileRobertMichon replied:
I'm leaning more towards making the RG delete blocking for now, that way we can test this and know if there are issues deleting RGs directly... otherwise, we risk leaking resources if there is a bug and we won't know. Thoughts?

devigned commented on cloud/errors.go:
```diff
@@ -18,12 +18,18 @@ package azure
 }
+
+// ResourceNotOwned parses the error to check if it's a resource not owned error.
+func ResourceNotOwned(err error) bool {
```
Did you specifically choose to write this method rather than using `errors.Is(err, ErrNotOwned)`?

Might be nice to use them uniformly like, `errors.Is(err, ResourceNotFound)` and `errors.Is(err, ResourceNotOwned)`.

CecileRobertMichon replied:
no, I didn't think of using errors.Is, I guess I just tried to be consistent with what was already there for not found. I can change it, not sure if it will work for NotFound too though (?)

devigned replied:
I think it would work for not found, but we would need to create an error type in our package which would wrap the inner not found error. Then implement `func (e *CloudError) Is(target error) bool {...}` where cloud error is that wrapping type.

Maybe this should be part of an error refactoring, rather than this PR.

Just wanted to bring it up, but I'm not super passionate about it.

devigned commented on cloud/services/groups/groups.go:
```diff
@@ -76,11 +76,11 @@ func (s *Service) Delete(ctx context.Context) error {
 }
 
-func (s *Service) isGroupManaged(ctx context.Context) (bool, error) {
+func (s *Service) IsGroupManaged(ctx context.Context) (bool, error) {
```
What is your thoughts on enforcing comments on public APIs?

CecileRobertMichon replied:
I generally like them :) sometimes I forget... Will add

devigned replied:
I was thinking about intro a PR to add to `make verify` to enforce public API comments. It might require some doc work to get us clean, but it would be nice to express that in our CI checks.

CecileRobertMichon replied:
I woudn't be opposed to that but it would require a lot of cleanup prior to being able to enforce since it would definitely fail with the codebase as-is. Do you mind opening an issue to track?

devigned reviewed (changes requested):
Fast delete is a huge improvement.

We probably want to intro an e2e test that will run fast delete and then a issue an apply with the same resources to ensure the cluster ends up in a good state.

CecileRobertMichon commented on controllers/helpers.go:
```diff
@@ -344,3 +346,16 @@ func GetMachinePoolByName(ctx context.Context, c client.Client, namespace, name
+		return true
+	}
+	grpSvc := groups.NewService(clusterScope)
+	managed, err := grpSvc.IsGroupManaged(ctx)
```
I'm having a really hard time testing this... since this makes an actual API call it needs to be mocked but the mock Client we have won't work for this since it's not of type azure.Service (azure.Service doesn't have a func `IsGroupManaged`)

devigned replied:
Is this exposed too high up the stack? Should the services understand they should skip delete if their cluster `IsGroupManaged` rather than the controller deciding to call delete on the underlying services?

CecileRobertMichon replied:
That'd be another way of doing it... but then that would require every single service to skip the delete instead of having the logic in one central place, which might be error-prone if new services are written in the future, no? Also would mean we would call every single service no matter what instead of deleting the RG right away (not a big deal but could be a slight de-optimization).

CecileRobertMichon replied:
The other thing is that we're not currently storing the `IsGroupManaged` value anywhere in spec (AzureCluster.Spec.ResourceGroup is a `string`) unlike the vnet for which we store the Tags, so every time we check we're reaching out to Azure right now. We shouldn't GET the RG tags for each resource during delete, so doing that would require a bigger change to start storing the RG tags. I actually had started doing that initially and was looking at either adding an annotation or a Status field but then that causes the problem of refreshing the tags when a delete is triggered to make sure we are up to date. All of this should be possible, I mostly just thinking "out loud".

CecileRobertMichon replied:
Also - we don't want to skip a service deletion if the resource belongs to something that isn't an AzureCluster that is getting individually deleted. For example, a public IP should be deleted if the AzureMachine it belongs to is getting deleted in the context of a scale down, but it shouldn't be deleted if the entire cluster is getting deleted. So I actually think the controller is the right layer too put this logic, not the services.

devigned replied:
Let's see how this plays for a bit. I'm not convinced the controller is where this should live. However, I do agree that each service should not need to call out to Azure to find out if the resource group is managed. 

I do think there is another way around each service calling out to Azure by making once call early in the reconcile cycle and caching the result in the cluster scope struct. I would imagine this should be added to the `ClusterDescriber` interface. Perhaps, the implementation there would do the memoization.

With that said, I don't want to hold up the PR. It's definitely an improvement. Let's just think on this one some more.

CecileRobertMichon commented:
/hold cancel

CecileRobertMichon commented:
> We probably want to intro an e2e test that will run fast delete and then a issue an apply with the same resources to ensure the cluster ends up in a good state.

For now I reverted the DeleteNoWait changes and decided to go for the more conservative approach of blocking on the actual RG delete before returning success. This is still a big improvement because deleting the RG directly takes almost 1/2 the time compared to deleting every resource one by one. Updated the PR description to reflect this.

devigned commented on cloud/services/groups/groups.go:
```diff
@@ -76,11 +76,13 @@ func (s *Service) Delete(ctx context.Context) error {
 	}
 
-	s.Scope.V(2).Info("successfully deleted resource group", "resource group", s.Scope.ResourceGroup())
+	s.Scope.V(2).Info("Deleting resource group. It might take a while for all resources to be completely gone...", "resource group", s.Scope.ResourceGroup())
```
Is this comment still accurate? Is this left over from the DeleteNoWait?

devigned replied:
Is this comment still accurate? Is this left over from the DeleteNoWait?

devigned commented on controllers/azurecluster_reconciler_test.go:
```diff
@@ -0,0 +1,131 @@
+					grp.Delete(context.TODO()).Return(nil))
+			},
+		},
+		"Resource Group is deletion fails": {
```
```suggestion
		"Resource Group delete fails": {
```

devigned reviewed (changes requested):
Couple little items

CecileRobertMichon commented:
`Test started today at 9:20 AM passed after 47m4s. (more info)`

We're definitely gaining a few minutes on e2e here, cc @mboersma

nader-ziada commented:
I think this is `lgtm` now, excited for the speed improvement this will bring

@devigned will let you decide since you did most of the review

devigned reviewed (approved):
/lgtm
/approve

This is a welcomed improvement in our delete times! I left one item to think about with regard to the controller logic vs services https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/938/files#r493637447.
